### PR TITLE
server: Separate txnRead from txnWrite

### PR DESCRIPTION
**Pull request by serathius**
cc @ahrtr @ptabor @jmhbnz @chaochn47 @fuweid

### CODE REVIEW

File: server/etcdserver/txn/txn.go [hunk 1]
```diff
@@ -278,7 +282,7 @@ func Txn(ctx context.Context, lg *zap.Logger, rt *pb.TxnRequest, txnModeWriteWit
 	return txnResp, trace, err
 }
 
-func checkTxn(ctx context.Context, txnWrite mvcc.TxnWrite, rt *pb.TxnRequest, isWrite bool, lessor lease.Lessor, txnPath []bool) error {
+func checkTxn(ctx context.Context, txnWrite mvcc.TxnRead, rt *pb.TxnRequest, isWrite bool, lessor lease.Lessor, txnPath []bool) error {
```
Comment (fuweid) [inline server/etcdserver/txn/txn.go:285]: we should rename the argu as well. s/txnWrite/txnRead/

File: server/etcdserver/txn/txn.go [hunk 2]
```diff
@@ -242,31 +242,35 @@ func Txn(ctx context.Context, lg *zap.Logger, rt *pb.TxnRequest, txnModeWriteWit
...
 	// ConcurrentReadTx to avoid extra overhead of copying buffer.
-	var txnWrite mvcc.TxnWrite
+	var mode mvcc.ReadTxMode
 	if isWrite && txnModeWriteWithSharedBuffer /*a.s.Cfg.ExperimentalTxnModeWriteWithSharedBuffer*/ {
-		txnWrite = mvcc.NewReadOnlyTxnWrite(kv.Read(mvcc.SharedBufReadTxMode, trace))
+		mode = mvcc.SharedBufReadTxMode
 	} else {
-		txnWrite = mvcc.NewReadOnlyTxnWrite(kv.Read(mvcc.ConcurrentReadTxMode, trace))
+		mode = mvcc.ConcurrentReadTxMode
 	}
```
Comment (chaochn47) [inline server/etcdserver/txn/txn.go:250]: nitpick on coding style.. 

```golang
mode := mvcc.ConcurrentReadTxMode
if isWrite && txnModeWriteWithSharedBuffer /*a.s.Cfg.ExperimentalTxnModeWriteWithSharedBuffer*/ {
    mode = mvcc.SharedBufReadTxMode
}
```

File: server/etcdserver/txn/txn.go (hunk 1 above)
Comment (chaochn47) [reply server/etcdserver/txn/txn.go:285]: +1

Comment (serathius) [reply server/etcdserver/txn/txn.go:285]: Done

File: server/etcdserver/txn/txn.go (hunk 2 above)
Comment (serathius) [reply server/etcdserver/txn/txn.go:250]: What style? Go style suggest avoiding `else` statements in separate main vs alternative code paths. However, this doesn't apply to normal conditions.

**Review by fuweid — APPROVED** [review]
LGTM

**Review by ahrtr — APPROVED** [review]

**Pull request merged**